evm/src/memory/memory_stark.rs

```diff
@@ -0,0 +1,577 @@
+    COUNTER, COUNTER_PERMUTED, IS_READ, NUM_REGISTERS, RANGE_CHECK, RANGE_CHECK_PERMUTED,
+    SEGMENT_FIRST_CHANGE, SORTED_ADDR_CONTEXT, SORTED_ADDR_SEGMENT, SORTED_ADDR_VIRTUAL,
+    SORTED_IS_READ, SORTED_TIMESTAMP, TIMESTAMP, VIRTUAL_FIRST_CHANGE,
+};
+use crate::stark::Stark;
+use crate::util::{permuted_cols, trace_rows_to_poly_values};
+use crate::vars::{StarkEvaluationTargets, StarkEvaluationVars};
+
+#[derive(Default)]
+pub struct TransactionMemory {
+    pub calls: Vec<ContractMemory>,
+}
+
+/// A virtual memory space specific to the current contract call.
+pub struct ContractMemory {
+    pub code: MemorySegment,
+    pub main: MemorySegment,
+    pub calldata: MemorySegment,
+    pub returndata: MemorySegment,
+}
+
+pub struct MemorySegment {
+    pub content: Vec<u8>,
+}
+
+pub(crate) const NUM_PUBLIC_INPUTS: usize = 0;
+
+#[derive(Copy, Clone)]
+pub struct MemoryStark<F, const D: usize> {
+    pub(crate) f: PhantomData<F>,
+}
+
+pub fn generate_random_memory_ops<F: RichField>(num_ops: usize) -> Vec<(F, F, F, F, F, [F; 8])> {
+    let mut memory_ops = Vec::new();
+
+    let mut rng = thread_rng();
+
+    let mut current_memory_values: HashMap<(F, F, F), [F; 8]> = HashMap::new();
+    for i in 0..num_ops {
+        let is_read = if i == 0 { false } else { rng.gen() };
+        let is_read_F = F::from_bool(is_read);
+
+        let (context, segment, virt, vals) = if is_read {
+            let written: Vec<_> = current_memory_values.keys().collect();
+            let &(context, segment, virt) = written[rng.gen_range(0..written.len())];
+            let &vals = current_memory_values
+                .get(&(context, segment, virt))
+                .unwrap();
+
+            (context, segment, virt, vals)
+        } else {
+            let context = F::from_canonical_usize(rng.gen_range(0..256));
+            let segment = F::from_canonical_usize(rng.gen_range(0..8));
+            let virt = F::from_canonical_usize(rng.gen_range(0..20));
+
+            let val: [u32; 8] = rng.gen();
+            let vals: [F; 8] = val.map(F::from_canonical_u32);
+
+            current_memory_values.insert((context, segment, virt), vals);
+
+            (context, segment, virt, vals)
+        };
+
+        let timestamp = F::from_canonical_usize(i);
+
+        memory_ops.push((timestamp, is_read_F, context, segment, virt, vals))
+    }
+
+    memory_ops
+}
+
+pub fn sort_memory_ops<F: RichField>(
+    timestamp: &[F],
+    is_read: &[F],
+    context: &[F],
+    segment: &[F],
+    virtuals: &[F],
+    values: &[[F; 8]],
+) -> (Vec<F>, Vec<F>, Vec<F>, Vec<F>, Vec<F>, Vec<[F; 8]>) {
+    let mut ops: Vec<(F, F, F, F, F, [F; 8])> = izip!(
+        timestamp.iter().cloned(),
+        is_read.iter().cloned(),
+        context.iter().cloned(),
+        segment.iter().cloned(),
+        virtuals.iter().cloned(),
+        values.iter().cloned(),
+    )
+    .collect();
+
+    ops.sort_by_key(|&(t, _, c, s, v, _)| {
```

Comment:
nit: I think we don't need the sorting to be stable here, so we can use `sort_unstable_by_key` which should be faster.